### PR TITLE
Models cache for CPU plugin

### PR DESCRIPTION
**Pull request by vladislav-volkov**
### Details:
- Custom NGraph nodes extenstion for CPU plugin
- Type relaxed opset serialization

### Tickets:
 - CVS-56555

### CODE REVIEW

File: inference-engine/src/inference_engine/ie_core.cpp [hunk 1]
```diff
@@ -224,7 +224,7 @@ class Core::Impl : public ICore {
     std::vector<IExtensionPtr> extensions;
 
     std::map<std::string, PluginDescriptor> pluginRegistry;
-    mutable std::mutex pluginsMutex;  // to lock parallel access to pluginRegistry and plugins
+    mutable std::recursive_mutex pluginsMutex;  // to lock parallel access to pluginRegistry and plugins
```
Comment (nosovmik) [inline inference-engine/src/inference_engine/ie_core.cpp:227]: @ilya-lavrenov, does inference engine call some plugin-specific code while locking its internal mutex? Maybe it would be good to fix this by not calling external code while locking mutexes, instead of using recursive ones?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_math_node.cpp [hunk 2]
```diff
@@ -16,12 +16,15 @@ using namespace InferenceEngine;
 
 bool MKLDNNMathNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& op, std::string& errorMessage) noexcept {
     try {
-        if (initializers.find(op->get_type_info()) == initializers.end()) {
+        if (std::find_if(initializers.begin(), initializers.end(),
```
Comment (nosovmik) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_math_node.cpp:19]: This piece of code is used in a lot of places. Is it possible to have some helper function like `find_type_info(initializers, op->get_type_info())`?

File: inference-engine/src/inference_engine/ie_core.cpp (hunk 1 above)
Comment (ilya-lavrenov) [reply inference-engine/src/inference_engine/ie_core.cpp:227]: @vladislav-volkov why do we need it?

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp [hunk 3]
```diff
@@ -777,6 +785,11 @@ std::shared_ptr<ngraph::Node> XmlDeserializer::createNode(
     const pugi::xml_node& node,
     const Blob::CPtr& weights,
     const V10Parser::GenericLayerParams& params) {
+    if (params.type == "GreaterEqual") {
```
Comment (nosovmik) [inline inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:788]: Is it a debug code?

**Comment by nosovmik** [comment]
As far as I understood, CPU plugin has extensions which can support additional operations. My suggestion is to add more 'caching tests' for CPU - currently it verifies only a very basic set of ngraph-functions for import/export (see tests/plugin/cpu/shared_test_instances/behavior/caching_tests.cpp), it would be good to add more 'functions' there for better coverage
BTW, these tests are currently failed on my Linux machine. Do you see the same behavior?

File: inference-engine/src/inference_engine/ie_core.cpp (hunk 1 above)
Comment (vladislav-volkov) [reply inference-engine/src/inference_engine/ie_core.cpp:227]: I add CPU plugin extensions from  MKLDNNPlugin::Engine::SetCore method and at this point, the mutex is already locked. The recursive mutex avoids deadlocks.

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 3 above)
Comment (vladislav-volkov) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:788]: Yes, I forgot to remove it.

File: inference-engine/src/inference_engine/ie_core.cpp (hunk 1 above)
Comment (ilya-lavrenov) [reply inference-engine/src/inference_engine/ie_core.cpp:227]: Discussed that we can remove `AddExtension` from plugin and after we don't need recursive mutex.
Plugin will take extensions from GetCore()->GetExtensions()

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_math_node.cpp (hunk 2 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_math_node.cpp]: fixed

File: inference-engine/src/inference_engine/ie_core.cpp (hunk 1 above)
Comment (vladislav-volkov) [reply inference-engine/src/inference_engine/ie_core.cpp:227]: PR 6493

File: cmake/developer_package/features.cmake [hunk 4]
```diff
@@ -58,7 +58,11 @@ ie_option (VERBOSE_BUILD "shows extra information about build" OFF)
 
 ie_option (ENABLE_UNSAFE_LOCATIONS "skip check for MD5 for dependency" OFF)
 
-ie_dependent_option (ENABLE_FUZZING "instrument build for fuzzing" OFF "CMAKE_CXX_COMPILER_ID MATCHES ^(Apple)?Clang$; NOT WIN32" OFF)
+if(CMAKE_CXX_COMPILER_ID MATCHES "^(Apple)?Clang$")
+    set(IS_CLANG ON)
+endif()
+
+ie_dependent_option (ENABLE_FUZZING "instrument build for fuzzing" OFF "IS_CLANG; NOT WIN32" OFF)
```
Comment (ilya-lavrenov) [inline cmake/developer_package/features.cmake:66]: fix is provided here https://github.com/openvinotoolkit/openvino/pull/6722 

Please, revert changes in this file

File: inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+NGRAPH_OP(StridedSlice, ngraph::op::v1)
+NGRAPH_OP(Subtract, ngraph::op::v1)
+NGRAPH_OP(Tan, ngraph::op::v0)
+NGRAPH_OP(Tanh, ngraph::op::v0)
+NGRAPH_OP(Tile, ngraph::op::v0)
+NGRAPH_OP(TopK, ngraph::op::v1)
+NGRAPH_OP(Transpose, ngraph::op::v1)
+NGRAPH_OP(Unsqueeze, ngraph::op::v0)
+NGRAPH_OP(VariadicSplit, ngraph::op::v1)
+NGRAPH_OP(Xor, ngraph::op::v0)
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp:185]: What is the binary size increase because of such duplication?

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 6]
```diff
@@ -601,5 +611,62 @@ QueryNetworkResult Engine::QueryNetwork(const CNNNetwork& network, const std::ma
...
+    networkModel.seekg(hdr.model_offset);
+    xmlString.resize(hdr.model_size);
+    networkModel.read(const_cast<char*>(xmlString.c_str()), hdr.model_size);
+
+    OV_ITT_SCOPE_SKIP(FIRST_INFERENCE, taskChain);
+
+    CNNNetwork cnnnetwork = GetCore()->ReadNetwork(xmlString, std::move(dataBlob), extensionManager->Extensions());
+
+    // TODO: Clarify the behavior of SetConfig method. Skip eng_config or not?
+    Config conf = engConfig;
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:655]: Can we remove this todo? Behavior is correct

File: cmake/developer_package/features.cmake (hunk 4 above)
Comment (vladislav-volkov) [reply cmake/developer_package/features.cmake]: fixed

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 6 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:663]: removed

File: inference-engine/src/inference_engine/ie_core.cpp (hunk 1 above)
Comment (vladislav-volkov) [reply inference-engine/src/inference_engine/ie_core.cpp]: recursive_mutex removed from this PR

File: inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp (hunk 5 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp:185]: We can't avoid increasing the binary size because TypeRelaxed class is template and we should register all type-relaxed opset in factory.

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 7]
```diff
@@ -610,5 +620,61 @@ QueryNetworkResult Engine::QueryNetwork(const CNNNetwork& network, const std::ma
...
 }
 
+InferenceEngine::IExecutableNetworkInternal::Ptr Engine::ImportNetwork(
+                                            const std::string &modelFileName,
+                                            const std::map<std::string, std::string> &config) {
+    std::fstream inputStream(modelFileName, std::ios::in | std::ios::binary);
+    if (inputStream.fail()) {
+        IE_THROW() << "Cannot open file to import model: " << modelFileName;
+    }
+    return ImportNetwork(inputStream, config);
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:630]: it already has default implementation, can we remove this override?

File: inference-engine/src/mkldnn_plugin/mkldnn_exec_network.h [hunk 8]
```diff
@@ -43,6 +43,10 @@ class MKLDNNExecNetwork: public InferenceEngine::ExecutableNetworkThreadSafeDefa
     INFERENCE_ENGINE_DEPRECATED("Use InferRequest::QueryState instead")
     std::vector<InferenceEngine::IVariableStateInternal::Ptr> QueryState() override;
 
+    void Export(const std::string &modelFileName) override;
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/mkldnn_exec_network.h:46]: it already has default implementation, can we remove this override?

File: inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp (hunk 5 above)
Comment (ilya-lavrenov) [reply inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp:185]: I see, but just FYI interesting to know what is the impact

File: inference-engine/src/inference_engine/ie_core.cpp [hunk 9]
```diff
@@ -485,6 +485,12 @@ class Core::Impl : public ICore, public std::enable_shared_from_this<ICore> {
         return details::ReadNetwork(model, weights, extensions);
     }
 
+    CNNNetwork ReadNetwork(const std::string& model, const Blob::CPtr& weights, const std::vector<IExtensionPtr>& exts) const override {
+        OV_ITT_SCOPE(FIRST_INFERENCE, itt::domains::IE_RT, "Core::Impl::ReadNetwork from memory with extensions");
+        return details::ReadNetwork(model, weights, exts);
+    }
```
Comment (ilya-lavrenov) [inline inference-engine/src/inference_engine/ie_core.cpp:491]: since we register CPU extensions in the ctor, do we need this custom overload then? I suppose Core already knows about CPU extensions..

File: inference-engine/src/mkldnn_plugin/mkldnn_exec_network.h (hunk 8 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_exec_network.h:46]: ok

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 7 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:630]: ok, i'll remove it

File: inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp (hunk 5 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp]: Difference between master and my branch is 733200 bytes.
master:                     libMKLDNNPlugin.so 29956328 bytes
dev-mkldnn-models-cache:    libMKLDNNPlugin.so 30689528 bytes

File: inference-engine/src/inference_engine/src/ie_core.cpp [hunk 10]
```diff
@@ -808,7 +811,10 @@ class CoreImpl : public InferenceEngine::ICore, public std::enable_shared_from_t
      */
     void AddExtension(const InferenceEngine::IExtensionPtr& extension) {
         std::lock_guard<std::mutex> lock(pluginsMutex);
+        AddExtensionUnsafe(extension);
+    }
 
+    void AddExtensionUnsafe(const InferenceEngine::IExtensionPtr& extension) const {
```
Comment (ilyachur) [inline inference-engine/src/inference_engine/src/ie_core.cpp:817]: Can we move this method to private?

File: inference-engine/src/inference_engine/src/ie_core.cpp [hunk 11]
```diff
@@ -6,6 +6,7 @@
 
 #include <sys/stat.h>
 
+#include <ie/details/ie_so_pointer.hpp>
```
Comment (ilyachur) [inline inference-engine/src/inference_engine/src/ie_core.cpp:9]: Why do you need to have this include?

File: ngraph/core/src/rt_info.cpp [hunk 12]
```diff
@@ -59,3 +59,34 @@ void ngraph::copy_runtime_info(const ngraph::NodeVector& from, ngraph::NodeVecto
         rtInfoTo = mergedInfo;
     }
 }
+
+static void merge_runtime_info(const ngraph::Node::RTMap& rt_info, std::shared_ptr<ngraph::Node> to) {
```
Comment (ilyachur) [inline ngraph/core/src/rt_info.cpp:63]: Why do we need these methods?

Comment (vladislav-volkov) [reply ngraph/core/src/rt_info.cpp:63]: Without this function we can lose some values from the RT info map. The copy_runtime_info is simply overwrite all RT info.

File: inference-engine/src/inference_engine/src/ie_core.cpp (hunk 11 above)
Comment (vladislav-volkov) [reply inference-engine/src/inference_engine/src/ie_core.cpp]: Removed

File: inference-engine/src/inference_engine/src/ie_core.cpp (hunk 10 above)
Comment (vladislav-volkov) [reply inference-engine/src/inference_engine/src/ie_core.cpp]: Moved

File: inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp (hunk 5 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_opset_extension.cpp]: After some optimizations, the overhead is reduced to 266272 bytes.

File: inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    return nullptr;
+}
+
+}  // namespace MKLDNNPlugin
+
+// Generate exported function
+IE_DEFINE_EXTENSION_CREATE_FUNCTION(MKLDNNPlugin::MKLDNNExtension)
+
+INFERENCE_EXTENSION_API(InferenceEngine::StatusCode)
+InferenceEngine::CreateExtension(InferenceEngine::IExtension*& ext, InferenceEngine::ResponseDesc* resp) noexcept {
```
Comment (ilyachur) [inline inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp:193]: If I understand right, it some internal extensions. I don't think that we need to create a new entry point to the plugin, users cannot load the plugin as a shared library with extensions.

File: inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+#undef NGRAPH_OP
+
+        return opset;
+    };
+
+    auto type_relaxed_opset = []() {
+        ngraph::OpSet opset;
+
+#define NGRAPH_OP(NAME, NAMESPACE) opset.insert<ngraph::op::TypeRelaxed<NAMESPACE::NAME>>();
+//        NGRAPH_OP(FullyConnectedNode, MKLDNNPlugin)
```
Comment (ilyachur) [inline inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp:48]: Why do we have a lot of commented operations?

File: inference-engine/src/mkldnn_plugin/mkldnn_serialize.cpp [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+                IE_THROW(NetworkNotRead) << "The input/output with name '" << name_attr.value() << "' not found";
+            }
+
+            it->second->setPrecision(Precision::FromStr(precision_attr.value()));
+            it->second->setLayout(layout_from_string(layout_attr.value()));
+        }
+    }
+};  // namespace
+
+CNNNetworkSerializer::CNNNetworkSerializer(std::ostream & ostream, MKLDNNExtensionManager::Ptr extensionManager)
```
Comment (ilyachur) [inline inference-engine/src/mkldnn_plugin/mkldnn_serialize.cpp:69]: Why cannot we reuse default serializer?

File: inference-engine/tests/functional/plugin/shared/include/behavior/core_integration.hpp [hunk 16]
```diff
@@ -431,8 +431,7 @@ TEST_P(IEClassBasicTestP, ImportNetworkThrows) {
     SKIP_IF_CURRENT_TEST_IS_DISABLED()
     Core ie = createCoreWithTemplate();
 
-    if (deviceName == CommonTestUtils::DEVICE_CPU ||
-        deviceName == CommonTestUtils::DEVICE_GPU) {
+    if (deviceName == CommonTestUtils::DEVICE_GPU) {
```
Comment (ilyachur) [inline inference-engine/tests/functional/plugin/shared/include/behavior/core_integration.hpp:434]: Why did you remove GPU?

File: inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp [hunk 17]
```diff
@@ -22,6 +23,7 @@
 #include <unordered_set>
 #include <utility>
 #include <cstring>
+#include <fstream>
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_exec_network.cpp:26]: removed

File: inference-engine/src/mkldnn_plugin/CMakeLists.txt [hunk 18]
```diff
@@ -44,6 +44,8 @@ target_link_libraries(${TARGET_NAME} PRIVATE mkldnn
                                              inference_engine_transformations
                                              inference_engine_lp_transformations)
 
+target_compile_definitions(${TARGET_NAME} PRIVATE IMPLEMENT_INFERENCE_EXTENSION_API)
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/CMakeLists.txt:47]: Since the plugin is implemented as an extension library, it should export the extension library interface. Therefore, this macro must be defined.

File: inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp (hunk 13 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp:193]: Unfortunately, the lifetime of the required extensions that are used for deserialization and the lifetime of the plugin are different. The plugin can be unloaded at any time. Therefore, the plugin is also added as an extension library if it provides the required extension interface.

File: inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp (hunk 14 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp:48]: I think it will be more convenient if we decide to add a new layer to type relaxed opset. We just need to uncomment the required layer.

File: inference-engine/src/mkldnn_plugin/mkldnn_serialize.cpp (hunk 15 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_serialize.cpp:69]: If you are talking about CNNNetwork::serialize, then unfortunately no. This implementation doesn't serialize CNN Network input and output precisions. A think, this behavior is incorrect for CNN Network serializer but I can't change existing implementation due to backward compatibility. Someone could use this serialization, and I can break his application.

File: inference-engine/tests/functional/plugin/shared/include/behavior/core_integration.hpp (hunk 16 above)
Comment (vladislav-volkov) [reply inference-engine/tests/functional/plugin/shared/include/behavior/core_integration.hpp:434]: I excluded the CPU from the condition because now the CPU plugin doesn't throw the required exception from ImportNetwork.

File: inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp (hunk 14 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_extension.cpp:48]: All comments removed

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 19]
```diff
@@ -617,5 +625,33 @@ QueryNetworkResult Engine::QueryNetwork(const CNNNetwork& network, const std::ma
     return res;
 }
 
+InferenceEngine::IExecutableNetworkInternal::Ptr Engine::ImportNetwork(std::istream& networkModel,
+                                            const std::map<std::string, std::string>& config) {
+    OV_ITT_SCOPE(FIRST_INFERENCE, itt::domains::MKLDNN_LT, "ImportNetwork");
+
+    CNNNetworkDeserializer deserializer(networkModel,
+        [this](const std::string& model, const Blob::CPtr& weights) {
+            return GetCore()->ReadNetwork(model, weights);
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:634]: Yes, sometimes this approach can work. After the plugin unloading, the registered extension will become invalid. Also, if I understand correctly, GetCore() can return a null pointer.

[the remark this answers is not in the document]
Comment (ilyachur) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:634]: Maybe we can unregister extensions in this case or have a clone of core object

File: ngraph/core/include/ngraph/rt_info.hpp [hunk 20]
```diff
@@ -20,6 +20,9 @@ void copy_runtime_info(std::shared_ptr<ngraph::Node> from, ngraph::NodeVector to
 NGRAPH_API
 void copy_runtime_info(const ngraph::NodeVector& from, std::shared_ptr<ngraph::Node> to);
 
+NGRAPH_API
+void copy_runtime_info(const ngraph::NodeVector& from, ngraph::NodeVector to);
+
```
Comment (ilya-lavrenov) [inline ngraph/core/include/ngraph/rt_info.hpp:25]: why do we need it? looks like the declaration below is the same

**Review by ilya-lavrenov — APPROVED** [review]

File: ngraph/core/include/ngraph/rt_info.hpp (hunk 20 above)
Comment (vladislav-volkov) [reply ngraph/core/include/ngraph/rt_info.hpp]: I think it was left after several mergers with the master.
Removed.

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:621]: The AddExtension registers extension in all loaded plugins and we have no guarantee that any plugin will not save the reference to the extension. We should implement RemoveExtension method for all plugins. I think the better solution, but not the best, is to implement the ReadNetwork method with an additional parameter in which we can pass a list of custom extensions. IMHO, changing the readers interface is not a good idea.

**Review by ilyachur — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]
